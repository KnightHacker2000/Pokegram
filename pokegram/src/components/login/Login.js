import PropTypes from 'prop-types';
import React, { useState } from 'react';
import Avatar from '@mui/material/Avatar';
import Button from '@mui/material/Button';
import Box from '@mui/material/Box';
import Container from '@mui/material/Container';
import Divider from '@mui/material/Divider';
import Typography from '@mui/material/Typography';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import { TextField } from '@mui/material';
import pokemon from '../../images/pikachu.jpg';
import userService from '../../services/userService';
import RootState from '../../models/rootState';

const theme = createTheme();

function Login(props) {
  const { parStates } = props;
  const [user, setUser] = useState({
    username: '',
    password: ''
  });
  const handleSubmit = async (event) => {
    event.preventDefault();
    // alert('current state is: ' + user.username + ' ' + user.password);
    // testing code for state manipulation
    // console.log(props);
    await userService.login();
    parStates.handleSetStates(true, true, user.username);
  };
  const handleUsername = (event) => {
    // console.log(event.target.value);
    const updateuser = {
      username: event.target.value,
      password: user.password
    };
    setUser(updateuser);
  };
  const handlePassword = (event) => {
    // console.log(event.target.value);
    const updateuser = {
      username: user.username,
      password: event.target.value
    };
    setUser(updateuser);
  };
  const handleSignup = (event) => {
    event.preventDefault();
    parStates.handleSetStates(false, false, -1);
  };
  return (
    <ThemeProvider theme={theme}>
      <Container maxWidth="xs">
        <Box sx={{
          marginTop: 8,
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center'
        }}
        >
          <Avatar alt="pikachu" sx={{ m: 1 }} src={pokemon} />
          <Typography component="h1" variant="h4">Wecome to Pokegram!</Typography>
        </Box>
        <Divider variant="middle" />
        <Box
          component="form"
          onSubmit={handleSubmit}
          noValidate
          sx={{
            marginTop: 20, my: 3, mx: 2, p: 4, border: '2px solid', borderRadius: '16px', boxShadow: 1, alignItems: 'center'
          }}
        >
          <Typography component="h1" variant="h6">Username:</Typography>
          <TextField margin="normal" required fullWidth id="username" label="Username" onChange={handleUsername} />
          <Typography component="h1" variant="h6">Password:</Typography>
          <TextField
            margin="normal"
            type="password"
            required
            fullWidth
            id="password"
            name="password"
            label="Password"
            value={user.password}
            onChange={handlePassword}
            error={user.password !== '' && user.password.length < 8}
            helperText={user.password !== '' && user.password.length < 8 ? 'Password should contains Minimum eight characters' : ' '}
          />
          <Button type="submit" fullWidth variant="contained" sx={{ mat: 3, mb: 2 }}>Sign In</Button>
        </Box>
        <Box sx={{
          my: 3, mx: 2, p: 1, marginTop: 8, border: 1, alignItems: 'center', display: 'flex'
        }}
        >
          <Typography component="h1" variant="h6" style={{ marginLeft: '16px', marginRight: '64px' }}>No Account?</Typography>
          <Button type="submit" variant="contained" sx={{ marginTop: '16px', mat: 3, mb: 2 }} onClick={handleSignup}>Sign Up</Button>
        </Box>
      </Container>
    </ThemeProvider>
  );
}
Login.propTypes = {
  parStates: PropTypes.instanceOf(RootState)
};

Login.defaultProps = {
  parStates: null
};
export default Login;
