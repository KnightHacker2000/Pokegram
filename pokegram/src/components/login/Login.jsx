import PropTypes from 'prop-types';
import React, { useEffect, useState } from 'react';
import Avatar from '@mui/material/Avatar';
import Button from '@mui/material/Button';
import Box from '@mui/material/Box';
import Container from '@mui/material/Container';
import Divider from '@mui/material/Divider';
import Typography from '@mui/material/Typography';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import { TextField } from '@mui/material';
import Swal from 'sweetalert2';
import withReactContent from 'sweetalert2-react-content';
import userService from '../../services/userService';
import RootState from '../../models/rootState';

const theme = createTheme();
const MySwal = withReactContent(Swal);

function Login(props) {
  const { parStates } = props;
  const [connected, setConnected] = useState(sessionStorage.getItem('app-token') !== null);
  const [user, setUser] = useState({
    username: sessionStorage.getItem('uid') == null ? 'test1' : sessionStorage.getItem('uid'),
    password: undefined
  });
  const [failed, setFailed] = useState(false);

  useEffect(() => {
    if (connected && sessionStorage.getItem('logout') === 'false') {
      parStates.handleSetStates(true, true, user.username);
    }
  });

  const handleUnlock = () => {
    setFailed(0);
  };

  const handleSubmit = async (event) => {
    event.preventDefault();
    // alert('current state is: ' + user.username + ' ' + user.password);
    // testing code for state manipulation
    // console.log(props);
    try {
      if (failed >= 5) {
        MySwal.fire('Please try again latter!');
        return;
      }
      if (
        connected || await userService.login(
          JSON.stringify({ id: user.username, password: user.password })
        )
      ) {
        setConnected(true);
        sessionStorage.setItem('logout', false);
        parStates.handleSetStates(true, true, user.username);
      }
      // console.log(res);
    } catch (err) {
      if (err.message && err.message === 'bad auth') {
        if (failed + 1 >= 5) {
          setFailed(failed + 1);
          setTimeout(handleUnlock, 50000);
        } else {
          MySwal.fire('Incorrect password or username!');
          setFailed(failed + 1);
        }
      }
    }
  };
  const handleUsername = (event) => {
    // console.log(event.target.value);
    const updateuser = {
      username: event.target.value,
      password: user.password
    };
    setUser(updateuser);
  };
  const handlePassword = (event) => {
    // console.log(event.target.value);
    const updateuser = {
      username: user.username,
      password: event.target.value
    };
    setUser(updateuser);
  };
  const handleSignup = (event) => {
    event.preventDefault();
    parStates.handleSetStates(false, false, -1);
  };
  return (
    <ThemeProvider theme={theme}>
      <Container maxWidth="xs">
        <Box sx={{
          marginTop: 8,
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center'
        }}
        >
          <Avatar alt="pikachu" sx={{ m: 1 }} src="" />
          <Typography component="h1" variant="h4">Wecome to Pokegram!</Typography>
        </Box>
        <Divider variant="middle" />
        <Box
          component="form"
          onSubmit={handleSubmit}
          noValidate
          sx={{
            marginTop: 20, my: 3, mx: 2, p: 4, border: '2px solid', borderRadius: '16px', boxShadow: 1, alignItems: 'center'
          }}
        >
          <Typography component="h1" variant="h6">Username:</Typography>
          <TextField margin="normal" required fullWidth id="username" label="Username" data-testid="test_username" onChange={handleUsername} />
          <Typography component="h1" variant="h6">Password:</Typography>
          <TextField
            data-testid="test_password"
            margin="normal"
            type="password"
            required
            fullWidth
            id="password"
            name="password"
            label="Password"
            value={user.password === undefined ? '' : user.password}
            onChange={handlePassword}
            error={user.password !== undefined && user.password !== '' && user.password.length < 8}
            helperText={user.password !== undefined && user.password !== '' && user.password.length < 8 ? 'Password should contains Minimum eight characters' : ' '}
          />
          <Button type="submit" fullWidth variant="contained" sx={{ mat: 3, mb: 2 }} data-testid="login_submit">Sign In</Button>
        </Box>
        <Box sx={{
          my: 3, mx: 2, p: 1, marginTop: 8, border: 1, alignItems: 'center', display: 'flex'
        }}
        >
          <Typography component="h1" variant="h6" style={{ marginLeft: '16px', marginRight: '64px' }}>No Account?</Typography>
          <Button type="submit" variant="contained" sx={{ marginTop: '16px', mat: 3, mb: 2 }} data-testid="login_signup" onClick={handleSignup}>Sign Up</Button>
        </Box>
      </Container>
    </ThemeProvider>
  );
}
Login.propTypes = {
  parStates: PropTypes.instanceOf(RootState)
};

Login.defaultProps = {
  parStates: null
};
export default Login;
