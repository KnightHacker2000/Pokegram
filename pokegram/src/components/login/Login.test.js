/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import Login from './Login';
import RootState from '../../models/rootState';
import '@testing-library/jest-dom/extend-expect';


const parentStates = new RootState(
  false,
  false,
  -1,
  (newLogin, newReg, newMyUID) => {
    console.log(newLogin, newReg, newMyUID);
  }
);

test('test Login snapshot', async () => {
  const loginComp = renderer.create(<Login parStates={parentStates}/>);
  const tree = loginComp.toJSON();
  // console.log(loginComp.toJSON());
  await expect(tree).toMatchSnapshot();
});

// test('test button click')
test('login button', async () => {
  render(<Login parStates={parentStates}/>);
  const submit = screen.getByTestId("login_submit");
  const username = screen.getByTestId("test_username");
  expect(username.value).toBe(undefined);
  console.log(submit);
  await fireEvent.click(submit);
  // const instance =
  // await userEvent.click(submit);
  //expect(submit.onSubmit).toHaveBeenCalled();
});

test('signup button', async () => {
  render(<Login parStates={parentStates} />);
  const signup= screen.getByTestId("login_signup");
  const password = screen.getByTestId("test_password");
  expect(password.value).toBe(undefined);
  console.log(signup);
  await fireEvent.click(signup);
  // const instance =
  // await userEvent.click(submit);
  //expect(submit.onSubmit).toHaveBeenCalled();
});