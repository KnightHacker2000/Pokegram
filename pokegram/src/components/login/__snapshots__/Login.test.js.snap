// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Login snapshot 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
>
  <div
    className="MuiBox-root css-binzgt"
  >
    <div
      className="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1xi8lsn-MuiAvatar-root"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
        data-testid="PersonIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
        />
      </svg>
    </div>
    <h1
      className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
    >
      Wecome to Pokegram!
    </h1>
  </div>
  <hr
    className="MuiDivider-root MuiDivider-middle css-ft4jij-MuiDivider-root"
  />
  <form
    className="MuiBox-root css-1wk4014"
    noValidate={true}
    onSubmit={[Function]}
  >
    <h1
      className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
    >
      Username:
    </h1>
    <div
      className="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      data-testid="test_username"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor="username"
        id="username-label"
      >
        Username
        <span
          aria-hidden={true}
          className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          id="username"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-yjsfm1"
          >
            <span>
              Username
               
              *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <h1
      className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
    >
      Password:
    </h1>
    <div
      className="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      data-testid="test_password"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor="password"
        id="password-label"
      >
        Password
        <span
          aria-hidden={true}
          className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-describedby="password-helper-text"
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          id="password"
          name="password"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="password"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-yjsfm1"
          >
            <span>
              Password
               
              *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        className="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-required css-1wc848c-MuiFormHelperText-root"
        id="password-helper-text"
      >
        <span
          className="notranslate"
        >
          ​
        </span>
      </p>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-axev6t-MuiButtonBase-root-MuiButton-root"
      data-testid="login_submit"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      Sign In
    </button>
  </form>
  <div
    className="MuiBox-root css-3i7hrm"
  >
    <h1
      className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
      style={
        {
          "marginLeft": "16px",
          "marginRight": "64px",
        }
      }
    >
      No Account?
    </h1>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1h553qu-MuiButtonBase-root-MuiButton-root"
      data-testid="login_signup"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      Sign Up
    </button>
  </div>
</div>
`;
